engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java

```diff
@@ -2279,95 +2279,151 @@ protected void migrate(final VMInstanceVO vm, final long srcHostId, final Deploy
 
-            if (_poolHostDao.findByPoolHost(targetPool.getId(), host.getId()) == null) {
-                throw new CloudRuntimeException(String.format("Cannot migrate the volume [%s] to the storage pool [%s] while migrating VM [%s] to target host [%s]. The host does not have access to the storage pool entered.", volume.getUuid(), targetPool.getUuid(), profile.getUuid(), host.getUuid()));
+            executeManagedStorageChecksWhenTargetStoragePoolProvided(currentPool, volume, targetPool);
+            if (_poolHostDao.findByPoolHost(targetPool.getId(), targetHost.getId()) == null) {
+                throw new CloudRuntimeException(
+                        String.format("Cannot migrate the volume [%s] to the storage pool [%s] while migrating VM [%s] to target host [%s]. The host does not have access to the storage pool entered.",
+                                volume.getUuid(), targetPool.getUuid(), profile.getUuid(), targetHost.getUuid()));
             }
             if (currentPool.getId() == targetPool.getId()) {
                 s_logger.info(String.format("The volume [%s] is already allocated in storage pool [%s].", volume.getUuid(), targetPool.getUuid()));
             }
             volumeToPoolObjectMap.put(volume, targetPool);
         }
-        return volumeToPoolObjectMap;
     }
 
     /**
-     * We create the default mapping of volumes and storage pools for the migration of the VM to the target host.
-     * If the current storage pool of one of the volumes is using local storage in the host, it then needs to be migrated to a local storage in the target host.
-     * Otherwise, we do not need to migrate, and the volume can be kept in its current storage pool.
+     * For each one of the volumes we will map it to a storage pool that is available via the target host.
+     * An exception is thrown if we cannot find a storage pool that is accessible in the target host to migrate the volume to.
      */
-    private Map<Volume, StoragePool> getDefaultMappingOfVolumesAndStoragePoolForMigration(VirtualMachineProfile profile, Host targetHost) {
-        Map<Volume, StoragePool> volumeToPoolObjectMap = new HashMap<Volume, StoragePool>();
-        List<VolumeVO> allVolumes = _volsDao.findUsableVolumesForInstance(profile.getId());
-        for (VolumeVO volume : allVolumes) {
+    private void createStoragePoolMappingsForVolumes(VirtualMachineProfile profile, Host targetHost, Map<Volume, StoragePool> volumeToPoolObjectMap, List<Volume> allVolumes) {
+        for (Volume volume : allVolumes) {
             StoragePoolVO currentPool = _storagePoolDao.findById(volume.getPoolId());
-            if (ScopeType.HOST.equals(currentPool.getScope())) {
-                createVolumeToStoragePoolMappingIfNeeded(profile, targetHost, volumeToPoolObjectMap, volume, currentPool);
+
+            executeManagedStorageChecksWhenTargetStoragePoolNotProvided(targetHost, currentPool, volume);
```

Comment:
This is the bit where I am confused. Let's say a VM has two  volumes. Volume A is in a managed storage at the cluster level, and volume B is in a local storage. Then, when I want to migrate this VM between hosts of the same cluster this method is triggered. 

However, an exception will be raised here since the VM needs to migrate a volume between local storage of hosts. Volume B is going to be mapped to a local storage at the target host, and volume A can be left where it is. The problem is that this validation here will raise an exception for volume A, because it is a cluster wide managed storage and we are not executing that checks between cluster IDs of the targetHost and current storage pool.

Do you understand what I am saying?

Reply:
Yes - take a look at the new code I proposed: https://github.com/mike-tutkowski/cloudstack/commit/bea1dad5b84bb80ba1f1297a69a1daea1c4ec736 .

It does not have any managed-storage verification logic in createStoragePoolMappingsForVolumes anymore.

I did have to add an if statement, though. If managed storage, then make sure the target host can see the managed storage in question (if it can't thrown an exception).

Reply:
So, to make sure the target host can "see" the current managed storage, isn't it a matter of checking if the the target host and the current storage pool of the volume (which is managed storage) are in the same cluster? 

Sorry I feel that we need to be able to discuss and just then exchange code. Otherwise, it is becoming trial and error which I really dislike.

Reply:
If the primary storage is zone wide, then it won't have a cluster ID.

Reply:
That is exactly what I understood before. What I do not understand is why we need those `executeManagedStorageChecksWhenTargetStoragePoolNotProvided` and `executeManagedStorageChecksWhenTargetStoragePoolProvided` instead of a single `executeManagedStorageChecks`.

Can you check at the last commit I pushed? Why would we need to divide that `executeManagedStorageChecks` into two other methods?